Optional faster version of the operator estimation UTs

msohaibalam opened the pull request:
Description
-----------

The unit tests for `operator_estimation` currently take ~4-5 mins to run. This PR improves on the run-time and reduces it down to ~35 secs for the entire test suite. This is essentially accomplished by introducing a seed for the random parameters in the various tests, which users can optionally switch off in favor of testing over a truly random choice of parameters by specifying the optional command line input `--use_seed`, e.g. running
```
pytest --use_seed=False pyquil/tests/test_operator_estimation.py
```
will run the longer version of the entire test suite, while something like
```
pytest --use_seed=False pyquil/tests/test_operator_estimation.py -k test_unitary_state_fidelity_readout_error
```
will run the longer version of the unit test `test_unitary_state_fidelity_readout_error`. However, the longer version comes at the cost of an increased number of experiments over which those results should be averaged in order to agree with expected results.

Additionally, this also unskips a previously skipped unit test, resolving issue #897 

Further, it bumps back the number of experiments (variously referred to as `num_simulations`, `num_expts`, or `runs` in the various unit tests) from 25 to 100. This number of experiments was previously determined to be consistent with the choice of `atol=2e-2` used throughout the test suite, and should therefore resolve issue #922 (among possibly others).

notmgsk commented on pyquil/tests/test_operator_estimation.py:
```diff
@@ -1758,7 +1915,7 @@ def test_corrected_statistics_2q_nontrivial_nonentangled_state(forest):
     # compare against simulated results
     np.testing.assert_allclose(result_expectation, expected_expectation, atol=2e-2)
-    np.testing.assert_allclose(result_std_err, expected_std_err, atol=2e-2)
+    np.testing.assert_allclose(result_std_err, expected_std_err, atol=3e-2)
```
This seems unrelated to the PR description

msohaibalam replied:
Reverted this back to original `atol` and picked appropriate seeds.

msohaibalam commented:
> How is the `use_seed` parameter configured? Do you feel these tests are convincing if they're only doing a single experiment?

The seeds are configured mostly by trial-and-error. Multiple experiments aren't as convincing as a single experiment, but a single experiment does provide evidence for the function's correctness.

notmgsk commented:
> > How is the `use_seed` parameter configured? Do you feel these tests are convincing if they're only doing a single experiment?
> 
> The seeds are configured mostly by trial-and-error. Multiple experiments aren't as convincing as a single experiment, but a single experiment does provide evidence for the function's correctness.

I mean, if I want to run the tests with the full 25 trials, do I manually have to call each test and set `use_seed=False`?

msohaibalam commented:
@notmgsk No, but I can (probably) add that in if that's convenient.

msohaibalam commented:
@notmgsk I've added in the ability to specify `use_seed` on the command line.

notmgsk commented:
255.59 seconds vs 45.31 seconds. Decent. I would, however, like to ensure that the *long* tests are enabled in semaphore. I suppose that will require changing `tox.ini`.

appleby commented on pyquil/tests/test_operator_estimation.py:
```diff
@@ -782,8 +848,8 @@ def test_exhaustive_symmetrization_2q(forest):
-# TODO: gh-897
-@pytest.mark.skip
+# # TODO: gh-897
+# @pytest.mark.skip
```
Why not just delete this rather than comment out?

msohaibalam replied:
Done.

msohaibalam commented:
@notmgsk and @appleby, thanks for the feedback! It might still be a few days before I could address your comments faithfully. But let me say this in the meantime: using a random seed or not in these tests is qualitatively not very different from the scenario where you do a (fair) coin-flip experiment a total of N times, and then calculate the frequency of observing heads. The average over several such experiments should of course be very close to 0.5, which is what the longer version of these tests check for. But in any single experiment, it may or may not be close to the same level of tolerance to 0.5. Are individual experiments with some fixed random seed still useful? It's a bit of a matter of perspective, but my perspective is yes, they are.

msohaibalam commented:
> Finally, there is a semi-standard pytest idiom of marking slow tests with a "slow" marker and allowing the user to skip those tests when they just want a quick sanity check to make sure nothing is horribly broken. Running slow tests can be [opt-in via a `--runslow` command line option](https://docs.pytest.org/en/latest/example/simple.html#control-skipping-of-tests-according-to-command-line-option), or [opt-out via the standard markers + the `-m 'not slow'` flag](https://docs.pytest.org/en/latest/mark.html#registering-marks). Providing a way to skip slow tests is maybe a wider question though, and not specific to this PR or the operator estimation tests.

@appleby I didn't want to skip any of these tests under any circumstances, since even their slow versions would protect against many conceivable breaking changes to the code. This is why I opted out of using the more standard `--runslow`, and chose `--use_seed` instead.

appleby commented:
> @notmgsk and @appleby, thanks for the feedback! It might still be a few days before I could address your comments faithfully. But let me say this in the meantime: using a random seed or not in these tests is qualitatively not very different from the scenario where you do a (fair) coin-flip experiment a total of N times, and then calculate the frequency of observing heads. The average over several such experiments should of course be very close to 0.5, which is what the longer version of these tests check for. But in any single experiment, it may or may not be close to the same level of tolerance to 0.5. Are individual experiments with some fixed random seed still useful? It's a bit of a matter of perspective, but my perspective is yes, they are.

Thanks for taking time to respond @msohaibalam. Don't worry, I know how code reviews tend to pile up :-).

I certainly didn't mean to suggest the fixed-seed versions provide *no* value, but based on your and @notmgsk's comments, it seems (as an outsider) like we are trading some test effectiveness for a lot of speed. The crux of my argument is that:

1.  test thoroughness trumps speed
2.  local tests should be as similar as possible to CI tests

Both of these come with a big "within reason" caveat. No one would suggest you need to test every combinatorial explosion of inputs for some function even if means tests take 10hrs to run, and I'm happy to leave it to the domain experts to determine whether these "rules" are bendable in this case. I'm just attempting to play devil's advocate here and push back a bit, in case there is a way to achieve our need for speed without weakening the test suite.

appleby commented:
> @appleby I didn't want to skip any of these tests under any circumstances, since even their slow versions would protect against many conceivable breaking changes to the code. This is why I opted out of using the more standard `--runslow`, and chose `--use_seed` instead.

I actually agree with you here. I only mentioned the `--runslow` option for completeness, but *if* we did go down the route of marking tests as "slow" my strong preference would be that "slow" tests get run by default, but the user can optionally skip them via `-m 'not slow'`.

This suggestion may seem at odds with my "test thoroughness trumps speed" claim, above, but the difference is that this option makes it explicit to the user the tradeoff they are making, i.e. skipping slow tests when they just want quick feedback while hacking on some unrealated feature. It should be clear in this context that you need to run the full test suite before opening a pull request. Whereas with the changes in this PR, it's not especially clear that the default is weaker/less thorough tests, and the `--use_seed` option is required to get the more thorough tests.

appleby commented:
One little caveat to my above rant, which I alluded to before but it's worth restating: if the changes in this PR are also intended to reduce test flakiness, as opposed to only speeding them up, then my objections to making fixed-seed the default are considerably softened.

msohaibalam commented:
@notmgsk @appleby Thanks again for your comments and feedback! I've made some changes to the code. Please give it another look when you can.

notmgsk commented:
Alright. So:

* Is it fair to say that we're confident, in both long-running and short-running mode, that the tests are doing their job? and
* If the above is true, why aren't we just using the short-mode by default?

msohaibalam commented on pyquil/tests/conftest.py:
```diff
@@ -189,3 +189,12 @@ def local_qvm_quilc():
+
+
+def pytest_addoption(parser):
+    parser.addoption("--use_seed", action="store", default=True, help="run some tests faster using random seed")
```
1. Changed `use_seed` to `use-seed`.
2. Picked option iii, but made the str to bool conversion function private
3. I'd rather stick with `--use-seed` than go for something like `--test-carefully` because the latter could be quite misleading (see my earlier comments about how even the fixed seed tests are *careful* enough), and the former makes very explicit what is being done. I would even argue that we use this flag in other places across the test suite where such a distinction between seeded vs averaged tests needs to be made.

Additionally, I've changed the default value to `False` so that Semaphore still runs the longer, and more complete versions of the tests, and if someone needs to iterate on `pyquil.operator_estimation` quickly enough, they can use the `use-seed` option.

msohaibalam commented:
@notmgsk The longer versions are set to default because (a) I thought you wanted the longer versions of the tests to run on Semaphore! :) , (b) I gave it some thought and actually agree that Semaphore should still run the longer versions.

As for (a), we briefly discussed this elsewhere some time ago. But loosely, you can think of the averaging effect as sort of applying the central limit theorem, you're drawing several samples i.i.d. from N copies of a random variable that has calculable expectation and variance, taking the average of those samples, which the CLT guarantees will converge to the true average. You can also do things like take the square root of the variance of those expectations, and see if they match up against the predicted standard error of the mean, as in `test_corrected_statistics_2q_nontrivial_nonentangled_state`.

Now of course, any single draw from that random variable would also likely be close to the average, and that's what the shorter version of the tests checks for.

The last reason is why the shorter, quicker tests are *reliable enough*, but the longer tests are more reliable (see also my coin flip example in one of the previous comments).

msohaibalam commented:
@notmgsk @appleby Ok, so since this is the 2nd time I mentioned an analogy with a "coin-flip experiment", here's a small code snippet that's quite analogous to what's being done here
```
import numpy as np


num_flips = 10   # number of flips in a coin-flip experiment
num_expts = 1000   # number of experiments
averages = []

for _ in range(num_expts):
    for _ in range(num_flips):
        draws = np.random.choice([0, 1], p=[0.5, 0.5], size=num_flips)
        avg = np.mean(draws)
        averages.append(avg)
```
If you now run
```
%matplotlib inline
import matplotlib.pyplot as plt

plt.hist(averages, bins=50)
plt.show()
```
you'll see a Gaussian looking curve centered around 0.5.
![image](https://user-images.githubusercontent.com/12733630/60994744-f27db780-a305-11e9-81ce-b4249bf3c498.png)

Moreover, if you picked a random draw from `averages`, i.e. drew `np.random.choice(averages)`, you'll find that the number is *quite likely* to be around 0.5, but not necessarily so. Indeed, for some sufficient choice of seed, that number would be pretty close to 0.5 (this is like the quicker, seeded version of the tests).

On the other hand, if you averaged over `averages`, i.e. calculate `np.mean(averages)`, you'll find that the number is pretty close to 0.5 (up to some tolerance) anyway.

appleby commented on pyquil/tests/conftest.py:
```diff
@@ -189,3 +189,22 @@ def local_qvm_quilc():
+
+
+def pytest_addoption(parser):
+    parser.addoption("--use-seed", action="store", type=_str_to_bool, default=False, help="run some tests faster using random seed")
```
Maybe change the help text to say "using a fixed random seed" since just "random seed" makes it sound like the seed is random (even though most people will know what you mean).

appleby commented:
> Moreover, if you picked a random draw from `averages`, i.e. drew `np.random.choice(averages)`, you'll find that the number is _quite likely_ to be around 0.5, but not necessarily so. Indeed, for some sufficient choice of seed, that number would be pretty close to 0.5 (this is like the quicker, seeded version of the tests).

Now this is where I'm getting well outside of the realm of Knowing What I'm Talking About, both in the general case of statistics and the specific case of what this code is testing. But by choosing a fixed seed, isn't true that you no longer actually know what the underlying distribution looks like? You might have changed the code in some way such that it's no longer Gaussian (for example), but just happens to return a value within the tolerance you're testing for your choice of seed. But I barely understand what a probability distribution is, so maybe I'm missing the point here :-).  This is why I defer to you as to whether making `--use-seed` the default makes sense.

msohaibalam commented:
@appleby I think the distribution itself is probably not as important as its mean and variance (in determining whether a random draw will be close to the mean/expected value). For the purposes of our tests, we really just care about the mean and variance.

notmgsk commented:
> @notmgsk The longer versions are set to default because (a) I thought you wanted the longer versions of the tests to run on Semaphore! :) , (b) I gave it some thought and actually agree that Semaphore should still run the longer versions.

I still feel that way *if* we need the longer tests at all, which is why I asked. And it seems we do want the longer tests. However, wasn't this prompted by Josh wanting specifically the semaphore tests too run faster? Anyway, if you're happy with the semaphore tests running the long version, so am I.

> As for (a), we briefly discussed this elsewhere some time ago. But loosely, you can think of the averaging effect as sort of applying the central limit theorem, you're drawing several samples i.i.d. from N copies of a random variable that has calculable expectation and variance, taking the average of those samples, which the CLT guarantees will converge to the true average. You can also do things like take the square root of the variance of those expectations, and see if they match up against the predicted standard error of the mean, as in `test_corrected_statistics_2q_nontrivial_nonentangled_state`.

👍 

> Now of course, any single draw from that random variable would also likely be close to the average, and that's what the shorter version of the tests checks for.
> 
> The last reason is why the shorter, quicker tests are _reliable enough_, but the longer tests are more reliable (see also my coin flip example in one of the previous comments).

appleby commented:
As a final parting thought, I'd just like to add that I'm leery here of being the least-informed person in the conversation who "wins" not by virtue of being right or having the best argument, but by virtue of being the most obstinate obstinate about repeating my (perhaps unfounded) objections. Which is why I keep harping on "lol. whatever u guys think". I don't want to be That Guy, if you know what I mean. So really: feel free to ignore my mild fixed-seed objections if they are not relevant.

msohaibalam commented:
@notmgsk I think @joshcombes might actually have meant being able to run the tests locally a lot quicker, which this PR would help do.
@appleby Feedback and code review is always helpful, so please don't shy away from it in the future either! :)
Thanks again to you both for the review!
Over to @karalekas now to give it a final look and merge it in.

karalekas commented:
Ok @msohaibalam I did my best to read through all the PR comments but I still have a few questions.

1. I'm still a bit confused about why the faster fixed seed version of the tests are less reliable. Could you give a summary of that?

2. If we must run the longer version, could we do it less? Could we run the long tests nightly and the short tests otherwise?

If this fast version of the tests is only used locally, and also is unreliable, I guess I'm unclear of what the advantage of having it around is. The CI pipeline for pyQuil is too slow, and that is what I would like for us to try to improve. Thoughts?

karalekas commented:
Also, changelog needs to be updated (re: the PR template).

notmgsk commented:
Gosh, we're really going round in circles now.

> Ok @msohaibalam I did my best to read through all the PR comments but I still have a few questions.
> 
> I'm still a bit confused about why the faster fixed seed version of the tests are less reliable. Could you give a summary of that?

I think this is the summary: https://github.com/rigetti/pyquil/pull/928#issuecomment-510173797

> 
> If we must run the longer version, could we do it less? Could we run the long tests nightly and the short tests otherwise?
> 
> If this fast version of the tests is only used locally, and also is unreliable, I guess I'm unclear of what the advantage of having it around is. The CI pipeline for pyQuil is too slow, and that is what I would like for us to try to improve. Thoughts?

I would like both to be improved. This solves at least one half of that, and since this PR has been open for too long now maybe fixing the other half (semaphore, gitlab) should be a separate issue.

msohaibalam commented:
@karalekas I'm not sure if I can add anything more useful beyond what I've already said in earlier comments. I'll just have to point you again to the coin-flip example I spelled out explicitly above to illustrate the idea that the faster, fixed random seed version of the tests are less reliable but still quite useful, while the slower, but more thorough, "averaging" version of the tests offer more air-tight protection against possibly more subtle breaking changes.

As someone who wants to contribute to the development of `pyquil.operator_estimation`, I would want to run the slower versions (at least locally) to quickly iterate on any code development. Before giving it final commits, or opening it up for review, I would want the longer, more thorough versions of the tests to run as a greater sanity check. It would be nice to have this be automatically done on Semaphore (hence, the current default is to run the longer versions there).

However, if the goal is to reduce time on Semaphore, then I'm more than happy to switch the default (especially if that will help close out this, frankly not very extensive, PR). In that case, I will personally just stick to the preferred protocol for local testing I stated above.

Also, I've updated the changelog. Please give it another look, and offer a decision on which version of the tests to run on Semaphore.

karalekas commented on docs/source/changes.rst:
```diff
@@ -8,6 +8,12 @@ Bugfixes:
 
+Improvements and Changes:
+
+- The test suite for `pyquil.operator_estimation` now has an (optional) faster version
```
let's just make sure it's clear that this is a pytest flag. i would do it but i cant push to your branch

```suggestion
- The test suite for `pyquil.operator_estimation` now has an (optional) faster version
  that uses fixed random seeds instead of averaging over several experiments. This can
  be enabled with the `--use-seed` command line input when running `pytest` (gh-928).
```

karalekas commented:
we'll leave the slower version on CI for now until someone thinks of an improvement and/or it becomes a hindrance to development

karalekas commented:
@msohaibalam Taking the suggestion via GH messed it up lol

karalekas commented on docs/source/changes.rst:
```diff
@@ -8,6 +8,14 @@ Bugfixes:
+- The test suite for `pyquil.operator_estimation` now has an (optional) faster version
+  that uses fixed random seeds instead of averaging over several experiments. This can
+  be enabled with the `--use-seed` command line option when running `pytest` (gh-928).
+  that uses fixed random seeds instead of averaging over several experiments, using the optional
```
```suggestion
```

karalekas commented on docs/source/changes.rst:
```diff
@@ -8,6 +8,14 @@ Bugfixes:
+  that uses fixed random seeds instead of averaging over several experiments. This can
+  be enabled with the `--use-seed` command line option when running `pytest` (gh-928).
+  that uses fixed random seeds instead of averaging over several experiments, using the optional
+  `--use-seed` command line input (gh-928).
```
```suggestion
```

msohaibalam commented:
@karalekas I tried taking the suggestion via GH again. If that fails again, I'll just edit and commit the old fashioned way.

karalekas commented:
Can you rebase? I'm worried the changelog wont automerge right. Put your bullet point below the existing ones in the "Improvements and Changes" section.

karalekas commented:
gonna wait for the build to complete and then merge
